src/Language/Fixpoint/Smt/Interface.hs

```diff
@@ -303,75 +264,76 @@ makeContextWithSEnv cfg f env = do
-makeProcess cfg
-  = do (hIn, hOut, _ ,pid) <- runInteractiveCommand $ smtCmd (solver cfg)
+makeProcess
+  :: Maybe Handle
+  -> ((LBS.ByteString -> IO ()) -> Process.Config)
+  -> IO (SMTLIB.Backends.Backend, IO ())
+makeProcess ctxLog cfg
+  = do handle     <- Process.new $ cfg $
+                       \s -> case ctxLog of
+                         Nothing -> return ()
+                         Just hLog ->
+                           LBS.hPutStrLn hLog $
+                             "OOPS, external process error: " <> s
+       let backend = Process.toBackend handle
+           p       = Process.process handle
+           hIn     = System.Process.Typed.getStdin p
+           hOut    = System.Process.Typed.getStdout p
+       hSetBuffering hOut $ BlockBuffering $ Just $ 1024 * 1024 * 64
+       hSetBuffering hIn $ BlockBuffering $ Just $ 1024 * 1024 * 64
+       return (backend, Process.kill handle)
+
+makeContext' :: Config -> Maybe Handle -> IO Context
+makeContext' cfg ctxLog
+  = do (backend, closeIO) <- case solver cfg of
+         Z3      ->
+           {- "z3 -smt2 -in"                   -}
+           {- "z3 -smtc SOFT_TIMEOUT=1000 -in" -}
+           {- "z3 -smtc -in MBQI=false"        -}
+           makeProcess ctxLog $ Process.Config "z3" ["-smt2", "-in"]
+         Z3mem   -> Conditional.Z3.makeZ3
+         Mathsat -> makeProcess ctxLog $ Process.Config "mathsat" ["-input=smt2"]
+         Cvc4    -> makeProcess ctxLog $
+                      Process.Config "cvc4" ["--incremental", "-L", "smtlib2"]
+       solver <- SMTLIB.Backends.initSolver SMTLIB.Backends.Queuing backend
        loud <- isLoud
-       hSetBuffering hOut $ BlockBuffering $ Just $ 1024*1024*64
-       hSetBuffering hIn $ BlockBuffering $ Just $ 1024*1024*64
-       -- See Note [Async SMT API]
-       queueTVar <- newTVarIO mempty
-       writerAsync <- async $ forever $ do
-         t <- atomically $ do
-           builder <- readTVar queueTVar
-           let t = Builder.toLazyText builder
-           when (LT.null t) retry
-           writeTVar queueTVar mempty
-           return t
-         LTIO.hPutStr hIn t
-         hFlush hIn
-       return Ctx { ctxPid     = pid
-                  , ctxIn      = hIn
-                  , ctxOut     = hOut
-                  , ctxLog     = Nothing
+       return Ctx { ctxSolver  = solver
+                  , ctxClose   = closeIO
+                  , ctxLog     = ctxLog
                   , ctxVerbose = loud
                   , ctxSymEnv  = mempty
-                  , ctxAsync   = writerAsync
-                  , ctxTVar    = queueTVar
                   }
 
 -- | Close file handles and wait for the solver process to terminate.
 cleanupContext :: Context -> IO ExitCode
-cleanupContext Ctx{..} = do
-  cancel ctxAsync
-  hCloseMe "ctxIn" ctxIn
-  hCloseMe "ctxOut"  ctxOut
+cleanupContext Ctx {..} = do
   maybe (return ()) (hCloseMe "ctxLog") ctxLog
-  waitForProcess ctxPid
+  ctxClose >> return ExitSuccess
```

Comment:
Should we also change the type of `cleanupContext` to `Context -> IO ()` ? Since it now only returns `ExitSuccess` and the return value doesn't seem to be used in the other modules.

Reply:
LGTM

Reply:
https://github.com/ucsd-progsys/liquid-fixpoint/pull/641/commits/afd8c755fcb1f83f33de27c78c3a1d994579c157